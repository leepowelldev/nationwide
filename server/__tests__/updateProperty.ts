import { gql } from 'apollo-server';
import { mockCollection } from './__utils__/mockCollections.js';
import setup from './__utils__/setup.js';

const query = gql`
  mutation ($id: ID!, $input: PropertyInput!) {
    updateProperty(id: $id, input: $input) {
      id
      address
      type
      bedrooms
    }
  }
`;

describe('Mutation: updateProperty', () => {
  test('should update property', async () => {
    const { server, collection } = setup(mockCollection);
    const input = {
      address: 'A house somewhere',
      type: 'HOUSE',
      bedrooms: 3,
    };

    // Check collection has default data
    expect(collection.get('1')?.address).toStrictEqual(
      mockCollection.get('1')?.address
    );
    expect(collection.get('1')?.bedrooms).toStrictEqual(
      mockCollection.get('1')?.bedrooms
    );

    const res = await server.executeOperation({
      query,
      variables: {
        id: '1',
        input,
      },
    });

    // Check collection has changed data
    expect(collection.get('1')?.address).toStrictEqual(input.address);
    expect(collection.get('1')?.bedrooms).toStrictEqual(input.bedrooms);
    expect(res).toMatchSnapshot();
  });

  test('should error when property does not exist', async () => {
    const { server, collection } = setup(mockCollection);
    const input = {
      address: 'A house somewhere',
      type: 'HOUSE',
      bedrooms: 3,
    };

    // Check collection has default data
    expect(collection.get('1')?.address).toStrictEqual(
      mockCollection.get('1')?.address
    );
    expect(collection.get('1')?.bedrooms).toStrictEqual(
      mockCollection.get('1')?.bedrooms
    );

    const res = await server.executeOperation({
      query,
      variables: {
        id: '2',
        input,
      },
    });

    // Check collection has not changed data
    expect(collection.get('1')?.address).toStrictEqual(
      mockCollection.get('1')?.address
    );
    expect(collection.get('1')?.bedrooms).toStrictEqual(
      mockCollection.get('1')?.bedrooms
    );
    expect(JSON.stringify(res.errors?.[0])).toMatchSnapshot();
    expect(res).toMatchSnapshot();
  });

  test('should error when property address is too short (yup schema)', async () => {
    const { server, collection } = setup(mockCollection);
    const input = {
      address: 'A hou',
      type: 'HOUSE',
      bedrooms: 3,
    };

    // Check collection has default data
    expect(collection.get('1')?.address).toStrictEqual(
      mockCollection.get('1')?.address
    );
    expect(collection.get('1')?.bedrooms).toStrictEqual(
      mockCollection.get('1')?.bedrooms
    );

    const res = await server.executeOperation({
      query,
      variables: {
        id: '1',
        input,
      },
    });

    // Check collection has not changed data
    expect(collection.get('1')?.address).toStrictEqual(
      mockCollection.get('1')?.address
    );
    expect(collection.get('1')?.bedrooms).toStrictEqual(
      mockCollection.get('1')?.bedrooms
    );
    expect(JSON.stringify(res.errors?.[0])).toMatchSnapshot();
    expect(res).toMatchSnapshot();
  });

  test('should error when property input is missing field (internal)', async () => {
    const { server, collection } = setup(mockCollection);
    const input = {
      address: 'A house somewhere',
      bedrooms: 3,
    };

    // Check collection has default data
    expect(collection.get('1')?.address).toStrictEqual(
      mockCollection.get('1')?.address
    );
    expect(collection.get('1')?.bedrooms).toStrictEqual(
      mockCollection.get('1')?.bedrooms
    );

    const res = await server.executeOperation({
      query,
      variables: {
        id: '1',
        input,
      },
    });

    // Check collection has not changed data
    expect(collection.get('1')?.address).not.toStrictEqual(input.address);
    expect(collection.get('1')?.bedrooms).not.toStrictEqual(input.bedrooms);
    expect(JSON.stringify(res.errors?.[0])).toMatchSnapshot();
    expect(res).toMatchSnapshot();
  });
});
