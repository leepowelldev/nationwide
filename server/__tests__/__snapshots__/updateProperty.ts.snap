// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation: updateProperty should error when property address is too short (yup schema) 1`] = `"{\\"message\\":\\"address must be at least 10 characters\\",\\"locations\\":[{\\"line\\":2,\\"column\\":3}],\\"path\\":[\\"updateProperty\\"],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\"}}"`;

exports[`Mutation: updateProperty should error when property address is too short (yup schema) 2`] = `
Object {
  "data": Object {
    "updateProperty": null,
  },
  "errors": Array [
    [GraphQLError: address must be at least 10 characters],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: updateProperty should error when property does not exist 1`] = `"{\\"message\\":\\"property with id 2 does not exist\\",\\"locations\\":[{\\"line\\":2,\\"column\\":3}],\\"path\\":[\\"updateProperty\\"],\\"extensions\\":{\\"code\\":\\"NOT_FOUND\\"}}"`;

exports[`Mutation: updateProperty should error when property does not exist 2`] = `
Object {
  "data": Object {
    "updateProperty": null,
  },
  "errors": Array [
    [GraphQLError: property with id 2 does not exist],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: updateProperty should error when property input is missing field (internal) 1`] = `"{\\"message\\":\\"Variable \\\\\\"$input\\\\\\" got invalid value { address: \\\\\\"A house somewhere\\\\\\", bedrooms: 3 }; Field \\\\\\"type\\\\\\" of required type \\\\\\"PropertyType!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":21}],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\"}}"`;

exports[`Mutation: updateProperty should error when property input is missing field (internal) 2`] = `
Object {
  "data": undefined,
  "errors": Array [
    [UserInputError: Variable "$input" got invalid value { address: "A house somewhere", bedrooms: 3 }; Field "type" of required type "PropertyType!" was not provided.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: updateProperty should update property 1`] = `
Object {
  "data": Object {
    "updateProperty": Object {
      "address": "A house somewhere",
      "bedrooms": 3,
      "id": "1",
      "type": "HOUSE",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
